### PR TITLE
Add ethereum/log events [#6601]

### LINKED ISSUE #6601: Add ethereum/log events
### User Story

As a developer, I want an event to access ethereum logs

### Description

*Type*: Feature

### PR DESCRIPTION
**Pull request by tbenr**
fixes #6601

### CODE REVIEW

**Comment by tbenr** [comment]
@jeluard I ended up supporting topics definition like that: every topic can be:
- direct topic, given by the extension itself (single value or vector)
  - examples:
    - `"0x29f31b934498b0deabbe211b01cc30eee6475abf0f0d553e7eb8ca71deeb24b3"`
    - `["0x29f31b934498b0deabbe211b01cc30eee6475abf0f0d553e7eb8ca71deeb24b3" "0xd7d1b6cffe566fcd240948f87999b06b8879e7fbfcf9d29362eda4395c1fa32e"]`
- by eventid and params
  - examples:
    - `{:event "Transfer" :params ["address" "address" "uint256"]}`
- by indexed param values
  - examples:
    - `{:type "uint256" :values ["125"]}`
    - `{:type "address" :values ["0x299b18709d4aacbda99048721448f65893a0c73a" "0x94eaa5fa6b313968b2abd6da375ef28077d95d53"]}`

putting all together, it is possible to use the new event as follows:

```
[ethereum/log {:fromBlock "0x4431D0"
               :toBlock "latest"
               :address ["0x2542F9D12B39a8F63C23A1f88948FA75303cBa3A" "0x869CD9C7D178254cD6ACF448737D7Eed2CA7E839"]
               :topics [{:event "Transfer" :params ["address" "address" "uint256"]}
                        {:type "address" :values ["0x299b18709d4aacbda99048721448f65893a0c73a" "0x94eaa5fa6b313968b2abd6da375ef28077d95d53"]}
                        nil
                        {:type "uint256" :values ["0x3fd"]}]                 
               :on-result [put-tbtest]}]
```

some considerations:
- the EDN for the event is not deeply checking the different options so it wont be able to verify at "installation time" the correctness of the different topics definition. It is not enforcing the "eventid and params" definition to be used in `topic[0]` only.
- for the "eventid and params" topic definition, I just realize it is not possible to use it in a vector of (when you want to search for different event logs in one call). I can implement it.
- result is taken as it is and sent to the `on-success` event.

**Comment by jeluard** [comment]
@tbenr Wow very nice work! 

> for the "eventid and params" topic definition, I just realize it is not possible to use it in a vector of (when you want to search for different event logs in one call). I can implement it.

That would be really cool!

Any chance you have some small extension that would demo it's usage?

File: src/status_im/extensions/ethereum.cljs [hunk 1]
```diff
@@ -72,3 +75,58 @@
...
+    (string/starts-with? type "fixed") :bytes
+    :else :bytes))
+
+(defn- values-topic-enc [type values]
+  (let [mapped-type (types-mapping type)]
+    (mapv #(str "0x" (abi-spec/enc {:type mapped-type :value %})) values)))
+
+(defn- parse-topic [t]
+  (cond
+    (nil? t) t ;; nil topic
```
Comment (jeluard) [inline src/status_im/extensions/ethereum.cljs]: Could be simplified with a or?

File: src/status_im/extensions/ethereum.cljs [hunk 2]
```diff
@@ -72,3 +75,58 @@
...
+         args {:jsonrpc "2.0"
+               :method constants/web3-get-logs
+               :params  [{:fromBlock (ensure-hex-bn fromBlock)
+                          :toBlock   (ensure-hex-bn toBlock)
+                          :address   address
+                          :topics    parsed-topics
+                          :blockhash blockhash}]}
+         payload (types/clj->json args)]
+     (status/call-private-rpc payload #(let [{:keys [error result]} (types/json->clj %1)
+                                             response (merge {:result result} (when error {:error error}))]
```
Comment (jeluard) [inline src/status_im/extensions/ethereum.cljs:130]: Would it make sense to have an `on-error` too?

File: src/status_im/extensions/ethereum.cljs (hunk 1 above)
Comment (tbenr) [reply src/status_im/extensions/ethereum.cljs]: what do you have in mind? I'm still new in clojure :)

Comment (jeluard) [reply src/status_im/extensions/ethereum.cljs]: Something like `(or (nil? t) (string? t) ..) t`

File: src/status_im/extensions/ethereum.cljs (hunk 2 above)
Comment (tbenr) [reply src/status_im/extensions/ethereum.cljs:130]: yes sure, but I was following the same bheaviour of the other events (`:extensions/ethereum-call`).
let's change them all?

Comment (jeluard) [reply src/status_im/extensions/ethereum.cljs:130]: Oh you are right my bad

**Comment by tbenr** [comment]
@jeluard thanks!

> Any chance you have some small extension that would demo it's usage?

I was actually a very basic\non-sense extension to test it. Let me try to clean it up.

> That would be really cool!

👍

**Comment by jeluard** [comment]
@tbenr  When you are done, you might be interested with those issues too : 

* https://github.com/status-im/status-react/issues/6855
* https://github.com/status-im/status-react/issues/6855

:)

**Review by flexsurfer — APPROVED** [review]

**Review by goranjovic — APPROVED** [review]

**Comment by tbenr** [comment]
@jeluard sure:) already applied for the ENS integration.
btw after rebasing to latest develop commits, i'm facing a bad java exception at runtime when i run the event (the event is executed, by the way)

`java.lang.IllegalArgumentException: Don't know how to create ISeq from: com.cognitect.transit.impl.TaggedValueImpl`

I have to check what's happening (if you have any hint...)

**Comment by jeluard** [comment]
@tbenr Doesn't ring a bell right now. I don't see how it could be related to your PR. Maybe the extension you are using?

**Comment by tbenr** [comment]
could be... in the weekend i'll figure out and make the final PR.

@jeluard btw the extension is: https://status-im.github.io/pluto/try.html?hash=QmPmydboqsmkuRg5DZZUEM7p4n5XNXtRgwwQc7tTa75fT4

**Comment by tbenr** [comment]
@jeluard a dummy extension containing some usage examples is here:

https://status-im.github.io/pluto/try.html?hash=QmfAWruXYfDzZVD59qWNL8DAxqsmpaFrsDeyd2gxJuuDRt

<blockquote><div><strong><a href="https://status-im.github.io/pluto/index.html">Pluto · A grammar for data manipulation</a></strong></div><div>A grammar for data manipulation</div></blockquote>

**Comment by jeluard** [comment]
Fantastic job @tbenr ! Can I just ask you to remove the `package-lock.json` from the commit before I merge?

**Comment by jeluard** [comment]
Oh also please rename `ethereum.log` to `ethereum.logs` :)

**Comment by tbenr** [comment]
@jeluard thanks!

last commit includes:
- `ethereum.log` to `ethereum.logs` renaming
- removed wrong file :)
- some tests

this is the updated sample extension: https://status-im.github.io/pluto/try.html?hash=Qmb7opEFGGTrjo6mt5DxjfZUJFLKEDFsTVsfb7LyTv7sxp


<blockquote><div><strong><a href="https://status-im.github.io/pluto/index.html">Pluto · A grammar for data manipulation</a></strong></div><div>A grammar for data manipulation</div></blockquote>

**Comment by jeluard** [comment]
@tbenr Awesome thanks!

**Comment by andytudhope** [comment]
Hey @tbenr please will you update your Eth address in Gitcoin so that I can pay out this bounty? Currently it will not let me do so.

**Comment by tbenr** [comment]
@andytudhope it was already in my account, in any case I updated it.

**Pull request merged**